src/v/redpanda/admin/security.cc

```diff
@@ -747,3 +746,41 @@ admin_server::update_role_members_handler(
     co_await throw_on_error(*req, err, model::controller_ntp);
     co_return ss::json::json_return_type(j_res);
 }
+
+ss::future<ss::json::json_return_type>
+admin_server::list_roles_handler(std::unique_ptr<ss::http::request> req) {
+    auto filter = req->get_query_param("filter");
+    auto user = req->get_query_param("principal");
+    auto principal_type = req->get_query_param("principal_type");
+
+    if (!principal_type.empty() && principal_type != "User") {
+        throw_role_exception(
+          role_errc::malformed_def,
+          fmt::format(
+            "Role membership reserved for user principals, got {{{}}}",
+            principal_type));
+    }
```

Comment:
nitpick: It would be nice if we could express more directly that `principal_type` takes a default value ("User") and feed that value (explicitly set or defaulted) into the predicate. More a limitation of the roles interface than anything I think.